src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -0,0 +1,974 @@
+import Plutarch.Prelude (
+    PAsData,
+    PBool,
+    PBuiltinList,
+    PBuiltinPair,
+    PByteString,
+    PEither (PLeft, PRight),
+    PEq ((#==)),
+    PInteger,
+    PIsData,
+    PIsListLike,
+    PLift,
+    PList,
+    PListLike (pcons, phead, pnil, pnull, ptail),
+    PMaybe (PJust, PNothing),
+    POrd,
+    PPair (PPair),
+    PPartialOrd ((#<), (#<=)),
+    PRational (PRational),
+    PString,
+    PUnit,
+    Type,
+    pconstant,
+    pdata,
+    pdcons,
+    pdenominator,
+    pdnil,
+    pfield,
+    pfix,
+    pfromData,
+    pfstBuiltin,
+    phoistAcyclic,
+    pif,
+    plet,
+    pnot,
+    pnumerator,
+    pshow,
+    psingleton,
+    psndBuiltin,
+    ptraceError,
+    (:-->),
+ )
+import Plutarch.Show (PShow)
+import Plutarch.TermCont (unTermCont)
+import Test.QuickCheck (
+    Arbitrary (arbitrary, shrink),
+    CoArbitrary (coarbitrary),
+    Function (function),
+    Gen,
+    Testable (property),
+    chooseInt,
+    elements,
+    frequency,
+    functionMap,
+    sized,
+    variant,
+    vectorOf,
+ )
+
+-- | @since 2.0.0
+instance Testable (TestableTerm PBool) where
+    property (TestableTerm t) = property (plift t)
+
+{- | TestableTerm is a wrapper for closed Plutarch term. This
+     abstraction allows Plutarch values can be generated via QuickCheck
+     Generator.
+
+     Hint: Typechecker is picky about how TestableTerm is constructed.
+     Meaning, TestableTerm throw error when it's composed.
+
+ @since 2.0.0
+-}
+data TestableTerm a = TestableTerm {unTestableTerm :: forall s. Term s a}
+
+-- | @since 2.0.0
+liftTestable ::
+    forall (a :: S -> Type) (b :: S -> Type).
+    (forall (s :: S). Term s a -> Term s b) ->
+    TestableTerm a ->
+    TestableTerm b
+liftTestable f (TestableTerm x) = TestableTerm $ f x
+
+-- | @since 2.0.0
+lift2Testable ::
+    forall (a :: S -> Type) (b :: S -> Type) (c :: S -> Type).
+    (forall (s :: S). Term s a -> Term s b -> Term s c) ->
+    TestableTerm a ->
+    TestableTerm b ->
+    TestableTerm c
+lift2Testable f (TestableTerm x) (TestableTerm y) = TestableTerm $ f x y
+
+-- | @since 2.0.0
+instance (forall (s :: S). Num (Term s a)) => Num (TestableTerm a) where
+    (+) = lift2Testable (+)
+    (*) = lift2Testable (*)
+    abs = liftTestable abs
+    negate = liftTestable negate
+    signum = liftTestable signum
+    fromInteger i = TestableTerm (fromInteger i :: Term s a)
+
+-- | @since 2.0.0
+instance (forall (s :: S). Eq (Term s a)) => Eq (TestableTerm a) where
+    (TestableTerm x) == (TestableTerm y) = x == y
+
+{- | For any Plutarch Type that have `PShow` instance, `Show` is
+     available as well. Unfortunately, for those that doesn't have
+     @PShow@, `forAllShow` with custom show function is required to
+     execute property check.
+
+     This only works with those that have PShow instance; it may cause
+     problems when dealing with Plutarch.Api datatypes which does not
+     have PShow instances.
+
+ @since 2.0.0
+-}
```

Comment:
"that has a 'PShow' instance, 'Show' is available as well."
"for those types that don't have 'PShow' instances, we have to use 'forAllShow' with a custom display function."

The second paragraph can go, as the first explains it well enough I feel.